### PR TITLE
Implement size limits for interpreter and intrinsic state

### PR DESCRIPTION
**Pull request by cmickeyb**

### CODE REVIEW

**Comment by cmickeyb** [comment]
@g2flyer good point about realloc. Let me take a look at that & see what can be done. it should not be difficult to add a function for realloc.

File: common/interpreter/gipsy_scheme/GipsyInterpreter.h [hunk 1]
```diff
@@ -25,15 +25,15 @@
 
 namespace pc = pdo::contracts;
 
-#define MAX_RESULT_SIZE 16000
-#define MAX_STATE_SIZE 64000
+#define MAX_RESULT_SIZE 1<<14
```
Comment (bvavala) [inline common/interpreter/gipsy_scheme/GipsyInterpreter.h]: MAX_RESULT_SIZE does not appear to be used anywhere.
Remove it or TBD?

File: common/interpreter/gipsy_scheme/safe_malloc.cpp [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+    // reallocate and update memory based on the new pointer
+    void *newptr = realloc(oldptr, request);
+    if (newptr == NULL)
+    {
+        SAFE_LOG1(PDO_LOG_WARNING, "gipsy memory allocation failed");
+        return newptr;
+    }
+
+    safe_malloc_map[(uint64_t)newptr] = request;
```
Comment (bvavala) [inline common/interpreter/gipsy_scheme/safe_malloc.cpp]: The pointer might already be present in the map due to a previous allocation.
1) it should be checked whether the pointer is present, or remove it first
2) if present, total_allocation should be adjusted (increased/decreased) according to the differential

File: common/packages/tinyscheme/scheme.c [hunk 3]
```diff
@@ -1175,6 +1175,7 @@ static char *store_string(scheme * sc, int len_str, const char *str,
     }
     if (str != 0) {
 	snprintf(q, len_str + 1, "%s", str);
+        q[len_str] = 0;
```
Comment (bvavala) [inline common/packages/tinyscheme/scheme.c]: Should we add 
/*
    The original source code has been modified to be used with Private Data Objects (PDOs).
*/
at the beginning of the file?

File: eservice/lib/libpdo_enclave/contract_worker.cpp [hunk 4]
```diff
@@ -35,11 +35,12 @@ void ContractWorker::InitializeInterpreter(void)
 
     if (current_state_ == INTERPRETER_DONE)
     {
-        if (interpreter_ != NULL) {
-            delete interpreter_;
+        if (interpreter_ == NULL) {
```
Comment (bvavala) [inline eservice/lib/libpdo_enclave/contract_worker.cpp]: Nice style opening a block at the end of the line -- it reminds me of someone! ;)

**Pull request merged**